components/Search.tsx

```diff
@@ -0,0 +1,353 @@
+import * as React from 'react'
+import { useRouter } from 'next/router'
+import { useQuery } from "@apollo/react-hooks"
+import { gql } from "apollo-boost"
+import { Alert, Button } from 'react-bootstrap'
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
+import { faSearch, faTimesCircle } from '@fortawesome/free-solid-svg-icons'
+import { faSquare, faCheckSquare } from '@fortawesome/free-regular-svg-icons'
+import Link from 'next/link'
+import ContentItem from "./ContentItem"
+import Error from "./Error"
+
+type Props = {
+
+}
+
+interface ContentNode {
+  node: ContentItem
+}
+
+interface ContentFacet {
+  id: string
+  title: string
+  count: number
+}
+
+interface PageInfo {
+  endCursor: string
+  hasNextPage: boolean
+}
+
+interface ContentQueryData {
+  works: {
+      __typename: String
+      nodes: ContentNode[]
+      pageInfo: PageInfo
+      published: ContentFacet[]
+      resourceTypes: ContentFacet[]
+      registrationAgencies: ContentFacet[]
+      totalCount: Number
+  },
+}
+
+interface ContentQueryVar {
+  query: string
+  cursor: string
+  published: string
+  resourceTypeId: string
+}
+
+export const CONTENT_GQL = gql`
+  query getContentQuery($query: String!, $cursor: String, $published: String, $resourceTypeId: String) {
+    works(first: 25, query: $query, after: $cursor, published: $published, resourceTypeId: $resourceTypeId) {
+      totalCount
+      pageInfo {
+        endCursor
+        hasNextPage
+      }
+      resourceTypes {
+        id
+        title
+        count
+      }
+      published {
+        id
+        title
+        count
+      }
+      registrationAgencies {
+        id
+        title
+        count
+      }
+      nodes {
+        id
+        doi
+        types {
+          resourceTypeGeneral
+          resourceType
+        }
+        titles {
+          title
+        }
+        creators {
+          id
+          name
+          givenName
+          familyName
+        }
+        descriptions {
+          description
+          descriptionType
+        }
+        publicationYear
+        publisher
+        version
+        citationCount
+        viewCount
+        downloadCount
+      }
+    }
+  }
+`
+
+export const Search: React.FunctionComponent<Props> = () => {
+  const router = useRouter()
+  // set initial searchQuery to query parameter in route
+  const [searchQuery, setSearchQuery] = React.useState(router.query.query as string || "")
```

Comment:
Using the router to manage the state is fine, though I do wonder if this shouldnt be seperated out a little more.
The searchQuery remains as just a simple string, this avoids needing to check for things like if searchQuery>0 and such as you only do anything on it when there is a data change, which should only be if there is data updates.

Maybe move the handling of the queryString to be seperate state and just sync the two as required, so if the queryString state gets updated you can update the searchQuery.
There is a newer hook built library for managing the query state that might help work with this see: https://www.npmjs.com/package/use-location-state

Reply:
Thanks @richardhallett I also think that this is the area that needs the most work. Will look at `use-location-state`.